sw/device/lib/base/bitfield.h

```diff
@@ -142,6 +142,98 @@ inline uint32_t bitfield_bit32_write(uint32_t bitfield,
                                 value ? 0x1u : 0x0u);
 }
 
+/**
+ * Find First Set Bit
+ *
+ * Returns one plus the index of the least-significant 1-bit of a 32-bit word.
+ *
+ * For instance, `ffs(field)` of the below 32-bit value returns `5`.
+ *
+ *     field:  0b00000000'00000000'11111111'00010000
+ *     index:   31                                 0
+ *
+ * @param bitfield Bitfield to find the first set bit in.
+ * @return One plus the index of the least-significant 1-bit of `bitfield`.
+ */
+BITFIELD_WARN_UNUSED_RESULT
+inline bitfield_bit32_index_t bitfield_ffs(uint32_t bitfield) {
+  return __builtin_ffs(bitfield);
```

Comment:
Do we want to do `ffs() - 1`, so that this is directly usable with functions that expect a bit32_index? This would make the "not present" value all-ones, which seems much more reasonable to me...?

Reply:
I am not sure, as doesn't the function just become `ctz` with that change? (but defined for zero).

Do we want to address that `ctz`/`clz` have undefined semantics for zero (especially as this is not the case for the B instruction set)

Reply:
Oh, that's a good point. Yeah, meh.

Re: UB on zero, would it be safer to definine ctz/clz to return 32 for zero?
```
return bitfield == 0 ? 32 : __builtin_ctz(bitfield);
```
This strikes me as being the "reasonable" semantic.

(I think, in general, we should not be in the business of propagating UB that the caller is probably going to need to do a runtime check for anyway.)

Reply:
Does (e.g.) `__builtin_ctz(0)` actually have **undefined** behaviour, or is it just implementation defined? This isn't actually part of the C standard. I see that the GCC online docs say "the result is undefined", not "evaluating this causes undefined behaviour".

I think the idea was that many processors support some sort of bit counting instruction, but different processors do different things on the weird corner cases. Rather than generate slow code (with a select or a branch), `__builtin_ctz` says "I'll do the right thing for the cases everyone agrees on, and I'll do *something* otherwise".

Since we control the toolchain and the processor, I should hope we can define that `__builtin_ctz(0)` means whatever the equivalent bit count instruction in the RV32B ISA gives you. Adding branches to the code seems a little bizarre to me.

Reply:
@rswarbrick Interestingly clang does have a hook for saying whether cltz of zero is undef, and on some platforms the answer is "no" (those platforms include ARM). On RISC-V (and x86) the answer is "yes" (even though this doesn't correspond to the B extension). 

The point about UB is the compiler currently gives no guarantees about what the behaviour is, or that the behaviour will be the same at different optimisation levels or on slightly different compiler versions. Remember also that we are trying to continue supporting both GCC and Clang, so avoiding divergence here is reasonable.

MCY's choice of 32 on zero matches the B specification, at least.

Reply:
Would we hope that clang/GCC could fold MCY's example code down to a single instruction with suitable optimisation settings?

Reply:
If we had the B extension, hopefully. I don't have an installation with B to hand that can help me with that though :/

Reply:
Correction, I do have an installation to hand. LLVM currently generates code with a branch if the C has a conditional. This is something we may be able to fix though, we haven't spent much time optimising code generation for the B extension.

Reply:
I think 99% of uses of clz and ctz will be with reggen symbols, which the compiler can just fold. If we're truly, genuinely desperate for the one-instruction version we can do
```c
#ifdef _whatever_the_rv32b_define_is_
int count;
asm volatile("ctz %1, %2" : "r"(count), "=r"(bitfield));
return count;
#else 
// ...
```
which I'm sure Sam won't find completely terrible. =)

I think this is the kind of thing that should be fixed in LLVM though. A single inlineable branch never hurt anyone.